src/shared/copy.c

```diff
@@ -794,6 +797,61 @@ static int prepare_nocow(int fdf, const char *from, int fdt, unsigned *chattr_ma
         return 0;
 }
 
+/* Copies fs-verity status.  May re-open fdt to do its job. */
+static int copy_fs_verity(int fdf, int *fdt) {
+        int r;
+
+        assert(fdf >= 0);
+        assert(fdt);
+        assert(*fdt >= 0);
+
+        r = fd_verify_regular(fdf);
+        if (r < 0)
+                return r;
+
+        struct fsverity_descriptor desc = {};
+        struct fsverity_read_metadata_arg read_arg = {
+                .metadata_type = FS_VERITY_METADATA_TYPE_DESCRIPTOR,
+                .buf_ptr = (uintptr_t) &desc,
+                .length = sizeof(desc),
+        };
+
+        r = ioctl(fdf, FS_IOC_READ_VERITY_METADATA, &read_arg);
+        if (r < 0) {
+                /* These mean that fs-verity isn't enabled on the file, filesystem, or kernel.  In all of
+                 * those cases, the correct thing to do is to do nothing at all.
+                 */
+                if (errno == ENODATA || ERRNO_IS_NOT_SUPPORTED(errno))
+                        return 0;
+                return -errno;
+        }
+
+        if (r < (int) sizeof desc)
+                /* This is really quite unexpected... */
+                return -EINVAL;
```

Comment:
we typically raise EINVAL if we determine data to be invalid that some code passed to us. But if we get back something from the kernel we should probably raise something else, or probably better assert() on it, because apparently we are not using the kernel apis right, and that'd be our bug.